File: app/src/main/java/it/gruppoinfor/home2work/services/SyncService.java
package it.gruppoinfor.home2work.services;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.IBinder;
import android.support.annotation.Nullable;
import android.util.Log;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import it.gruppoinfor.home2work.App;
import it.gruppoinfor.home2work.database.RoutePointEntity;
import it.gruppoinfor.home2work.database.RoutePointRepo;
import it.gruppoinfor.home2work.utils.SessionManager;
import it.gruppoinfor.home2work.utils.UserPrefs;
import it.gruppoinfor.home2workapi.model.Location;


public class SyncService extends Service {

    private final String TAG = "SYNC_SERVICE";

    private List<Location> locations = new ArrayList<>();

    @Override
    public void onCreate() {
        super.onCreate();

        UserPrefs.init(this);

        SessionManager sessionManager = new SessionManager(this, new SessionManager.SessionManagerCallback() {
            @Override
            public void onNoSession() {

            }

            @Override
            public void onValidSession() {
                if (getConnectivityType(SyncService.this) == ConnectivityManager.TYPE_WIFI || UserPrefs.syncWithData) {
                    sync();
                }
            }

            @Override
            public void onExpiredToken() {

            }

            @Override
            public void onError(Throwable throwable) {

            }
        });
        sessionManager.loadSession();

    }

    public void sync() {
        RoutePointRepo.getAllUserLocations()
                .subscribe(routePointEntities -> {

                    for (RoutePointEntity routePointEntity : routePointEntities) {

                        Location location = new Location();
                        location.setLatLng(routePointEntity.getLatLng());
                        location.setDate(new Date(routePointEntity.getTimestamp() * 1000));
                        locations.add(location);

                    }

                    if (locations.size() > 0) syncRoutePoints(locations);

                });

    }

    private void syncRoutePoints(final List<Location> locationList) {

        App.home2WorkClient.uploadLocation(locationList,
                locations -> RoutePointRepo.deleteAllUserLocations().subscribe(),
                e -> Log.e(TAG, "Sincronizzazione fallita", new Throwable(e)));

    }

    private int getConnectivityType(Context context) {
        ConnectivityManager cm = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo activeNetwork = cm.getActiveNetworkInfo();
        if (activeNetwork != null) { // connected to the internet
            return activeNetwork.getType();
        } else {
            return -1;
        }
    }

    private boolean canSync() {
        if (getConnectivityType(this) != ConnectivityManager.TYPE_WIFI) {
            if (!UserPrefs.syncWithData || getConnectivityType(this) != ConnectivityManager.TYPE_MOBILE)
                return false;
        }
        return true;
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
}

File: App/src/main/java/it/gruppoinfor/home2work/services/SyncService.kt
package it.gruppoinfor.home2work.services

import android.app.Service
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import android.os.IBinder
import android.util.Log
import com.crashlytics.android.answers.Answers
import com.crashlytics.android.answers.CustomEvent
import com.google.android.gms.tasks.OnFailureListener
import com.google.android.gms.tasks.OnSuccessListener
import com.pixplicity.easyprefs.library.Prefs
import it.gruppoinfor.home2work.App
import it.gruppoinfor.home2work.model.UserLocation
import it.gruppoinfor.home2work.model.UserLocation_
import it.gruppoinfor.home2work.user.SessionManager
import it.gruppoinfor.home2work.utils.Const
import it.gruppoinfor.home2workapi.HomeToWorkClient
import it.gruppoinfor.home2workapi.location.Location
import java.util.*


class SyncService : Service() {

    private val routeLocations = ArrayList<Location>()

    override fun onCreate() {
        super.onCreate()

        SessionManager.loadSession(this, object : SessionManager.SessionCallback {
            override fun onValidSession() {
                if (canSync()) sync()
            }

            override fun onInvalidSession(code: Int, throwable: Throwable?) {
                throwable?.printStackTrace()
            }
        })

    }

    fun sync() {

        val userLocationBox = App.boxStore.boxFor(UserLocation::class.java)
        val userLocations = userLocationBox.query().equal(UserLocation_.userId, HomeToWorkClient.user!!.id).build().find()

        userLocations.forEach {
            val routeLocation = it.gruppoinfor.home2workapi.location.Location()
            routeLocation.latLng = it.latLng
            routeLocation.date = Date(it.timestamp * 1000)
            routeLocations.add(routeLocation)
        }

        if (routeLocations.size > 0) syncRoutePoints(routeLocations)


    }

    private fun syncRoutePoints(locationList: List<Location>) {

        HomeToWorkClient.uploadLocations(locationList,
                OnSuccessListener {

                    Answers.getInstance().logCustom(CustomEvent("Sincronizzazione posizioni"))

                    val userLocationBox = App.boxStore.boxFor(UserLocation::class.java)
                    userLocationBox.query().equal(UserLocation_.userId, HomeToWorkClient.user!!.id).build().remove()

                },
                OnFailureListener { e -> Log.e(this::class.java.name, "Sincronizzazione fallita", Throwable(e)) })

    }

    private fun getConnectivityType(context: Context): Int {

        val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager?
        if (cm != null) {
            val activeNetwork = cm.activeNetworkInfo
            if (activeNetwork != null) { // connected to the internet
                return activeNetwork.type
            }
        }

        return -1
    }

    private fun canSync(): Boolean {

        val wifiEnabled = getConnectivityType(this@SyncService) == ConnectivityManager.TYPE_WIFI

        return wifiEnabled || Prefs.getBoolean(Const.PREFS_SYNC_WITH_DATA, true)
    }

    override fun onBind(intent: Intent): IBinder? {

        return null
    }
}
